### ISSUE TITLE
State machine for characters/Trait System

### ISSUE THREAD

**Issue by IllustratedMan-code** [quaternary-game/quaternary#3]
# The idea
Characters need some way of determining their behavior, and some behaviors should be determined by their traits. To achieve this, all of the characters will be state machines. Luckily, we've all taken (or are taking) automata, so we should all be familiar with this design pattern. 

States (and sets of states) are "behaviors". When the organism is in a particular state, it exhibits a behavior like "moving" or "searching for food". When exposed to some input (a valid transition), the organism will transition to another state and exhibit the new behavior. A simple character behavior might look like this:
![image](https://github.com/quaternary-game/quaternary/assets/60514384/a54492cb-3648-4077-b04a-5cf7daf2bd4c)

It is technically an NFA with a potentially infinite "alphabet". I've written the following code to model a simple behavior I call "move randomly".  FYI the language is "`gdscript`" for github markdown syntax highlighting.
```gdscript
extends CharacterBody2D


var SPEED = 80


# States
class State:
	func function():
		pass
	func transition():
		return ""

		
# State handler (the machine)
class StateMachine:
	var states: Dictionary # {statename: stateobject}
	var current_state: State

	func _init(_states):
		states = _states
		current_state = states[states.keys()[0]]
	func transition():
		current_state.function()
		current_state = states[current_state.transition()]
		
	
func load_trait(trait_name):
	Traits.append(trait_name)
	
func _process(delta):
	pass

class random_direction extends State:
	
	
	var global_node
	
	func _init(global_self):
		global_node = global_self
		
	func function():
		var rng = RandomNumberGenerator.new()
		var new_direction = Vector2(1, 0).rotated(rng.randf() * 2 *PI)
		global_node.velocity = new_direction * global_node.SPEED

	
	func transition():
		return "move random"
		

class move_random extends State:
	var global_node
	var timer = Timer.new()
	var ready: bool = false
	func _init(global_self):
		global_node = global_self
		global_node.add_child(timer)
		timer.set_paused(true)
		timer.set_one_shot(true)
	func function():
		var rng = RandomNumberGenerator.new()
		if timer.is_paused():
			timer.set_paused(false)
			timer.start(rng.randf_range(0.01, 5))
		if timer.is_stopped():
			timer.set_paused(true)
			ready = true
			return
		global_node.move_and_slide()
	func transition():
		if ready:
			ready = false
			return "random direction"
		else:
			return "move random"


var m = StateMachine.new({"random direction":random_direction.new(self), "move random": move_random.new(self)})

func _physics_process(delta):
	m.transition()
	
```

![image](https://github.com/quaternary-game/quaternary/assets/60514384/16c3d84d-9721-47be-8d02-aad59a9640ab)

# Traits: Where it gets stinky

We need to figure out a dynamic way of attaching new states via transitions. When a trait is acquired, new behaviors should be unlocked. Gaining the "Herbivore" trait might change the "looking for food" behavior. Potentially, we might need multiple state machines working in tandem within a single organism (ie multiple sensory behaviors acting in parallel). Looking at the [signals](https://docs.godotengine.org/en/stable/getting_started/step_by_step/signals.html) documentation, it might be easy to manage the states by connecting signals, rather than the way I currently have it implemented.

**Comment by IllustratedMan-code** [comment]
# Style guide for traits
- If possible, there should be a visual cue for each trait on the organism
- Needs to have a significant effect on the organism's survival (quantifiable)
- Traits should have a cost and a benefit to the organism as well. This could be quantified by calories or something.

# Ideas for possible traits
- Eyes (directional vision)
- Feet 
- Fat
- Sweating
- Skin color?
- Herbivore/Carnivore/Omnivore (different mouths?)
- Reproduction Speed
- Growth rate
- Photosynthesis
- Nose/Ears (good for navigating caves/increases field of view)
- Secondary sex characteristics (help identify mates)

# Behavioral traits
These aren't as quantifiable, nor as visually interesting, but could be interesting. I'm not really sure how to balance these either.
- Memory: number of other cells/areas/organisms a cell can remember 
- Pathfinding 
- Solitary/Group behavior
- Spinning (that's a good trick), could help to make eyes useful without ears or nose

**Comment by john-whiting** [comment]
This is an excellent breakdown of the trait behaviors. It could also reach the point where we have a tool that lets us graphically lay out the state machines, and those settings configure the game.

**Comment by IllustratedMan-code** [comment]
ECS @john-whiting

**Comment by IllustratedMan-code** [comment]
Since we decided to go for an ECS system designed by John, closing this. @john-whiting Maybe start a new issue with your progress and ideas behind architecture?
